### ISSUE TITLE
accuray always drops when testing Stacked Denoising Autoencoder

### ISSUE THREAD

**Issue by yzsatgithub** [blackecho/Deep-Learning-TensorFlow#5]
tried several sets of hyperparameters:
first tried those parameters in your documentation:
```
python command_line/run_stacked_autoencoder_supervised.py --layers 1024,784,512,256 --batch_size 25 --num_epochs 5 --verbose 1 --corr_type masking --corr_frac 0.0 --finetune_learning_rate 0.002 --finetune_num_epochs 25 --opt momentum --momentum 0.9 --learning_rate 0.05 --enc_act_func sigmoid --finetune_act_func relu --dropout 0.7
```
```
Accuracy at step 0: 0.1502
Accuracy at step 1: 0.1126
Accuracy at step 2: 0.1126
Accuracy at step 3: 0.1126
Accuracy at step 4: 0.1126
Accuracy at step 5: 0.1126
Accuracy at step 6: 0.1126
Accuracy at step 7: 0.1126
Accuracy at step 8: 0.1126
Accuracy at step 9: 0.1126
Accuracy at step 10: 0.1126
Accuracy at step 11: 0.1126
Accuracy at step 12: 0.1126
Accuracy at step 13: 0.1126
```
changed the finetuning learning rate:
```
python command_line/run_stacked_autoencoder_supervised.py --layers 1024,784,512,256 --batch_size 25 --num_epochs 5 --verbose 1 --corr_type masking --corr_frac 0.0 --finetune_learning_rate 0.0002 --finetune_num_epochs 25 --opt momentum --momentum 0.9 --learning_rate 0.05 --enc_act_func sigmoid --finetune_act_func relu --dropout 0.7
```
```
Accuracy at step 0: 0.1002
Accuracy at step 1: 0.1002
Accuracy at step 2: 0.0868
Accuracy at step 3: 0.0868
Accuracy at step 4: 0.0868
Accuracy at step 5: 0.0868
Accuracy at step 6: 0.0868
Accuracy at step 7: 0.0868
Accuracy at step 8: 0.0868
Accuracy at step 9: 0.0868
Accuracy at step 10: 0.0868
Accuracy at step 11: 0.0868
Accuracy at step 12: 0.0868
Accuracy at step 13: 0.0868
Accuracy at step 14: 0.0868
Accuracy at step 15: 0.0868

```
changed dropout rates:
```
python command_line/run_stacked_autoencoder_supervised.py --layers 1024,784,512,256 --batch_size 25 --num_epochs 5 --verbose 1 --corr_type masking --corr_frac 0.0 --finetune_learning_rate 0.0002 --finetune_num_epochs 25 --opt momentum --momentum 0.9 --learning_rate 0.05 --enc_act_func sigmoid --finetune_act_func relu --dropout 0.9
```
```
Accuracy at step 0: 0.094
Accuracy at step 1: 0.0992
Accuracy at step 2: 0.11
Accuracy at step 3: 0.11
Accuracy at step 4: 0.11
Accuracy at step 5: 0.11
Accuracy at step 6: 0.11
Accuracy at step 7: 0.11
Accuracy at step 8: 0.11
Accuracy at step 9: 0.11
Accuracy at step 10: 0.11
Accuracy at step 11: 0.11
Accuracy at step 12: 0.11
Accuracy at step 13: 0.11
Accuracy at step 14: 0.11
```
changed network structure:
```
python command_line/run_stacked_autoencoder_supervised.py --layers 512,128 --batch_size 25 --num_epochs 5 --verbose 1 --corr_type masking --corr_frac 0.0 --finetune_learning_rate 0.002 --finetune_num_epochs 25 --opt momentum --momentum 0.9 --learning_rate 0.05 --enc_act_func sigmoid --finetune_act_func relu --dropout 1.0
```
```
Accuracy at step 0: 0.1804
Accuracy at step 1: 0.1986
Accuracy at step 2: 0.1786
Accuracy at step 3: 0.1614
Accuracy at step 4: 0.1468
Accuracy at step 5: 0.1346
Accuracy at step 6: 0.1192
Accuracy at step 7: 0.1066
Accuracy at step 8: 0.101
Accuracy at step 9: 0.1002
Accuracy at step 10: 0.1002
```

Has this something to do with network structure I chose?
By the way, cost in the pretraining process are decreasing which is reasonable.

**Comment by blackecho** [comment]
I just found an error in the softmax layer initialization. I didn't specify the standard deviation in the weights initialization, and the default is 1, which is too big. I just changed it to 0.01, now it should work better. Try first with a smaller network, like the 512,128 you tried!

**Comment by blackecho** [comment]
Try with this:

    python command_line/run_stacked_autoencoder_supervised.py --layers 1024,256 --verbose 1 --finetune_learning_rate 0.001 --finetune_opt momentum --momentum 0.8 --finetune_num_epochs 10 --finetune_act_func tanh --dropout 0.7 --finetune_batch_size 64 --finetune_loss_func mean_squared --do_pretrain true --seed 0 --batch_size 64  --num_epochs 3 --enc_act_func sigmoid --dec_act_func sigmoid --loss_func cross_entropy

You should get an accuracy of about 86%.

**Comment by yzsatgithub** [comment]
@blackecho  The latest code works really well ~  classifying MNIST using Stacked Denoising Autoencoder ends with an accuracy of 0.9374 for validation set  after 25 epochs and 0.93400 for test set. :）

P.S. saw you update your document :+1:   :)

**Comment by blackecho** [comment]
Perfect, I'll close the issue :) :+1:
